pandas/core/arrays/masked.py

```diff
@@ -870,7 +870,9 @@ def take(
         # we only fill where the indexer is null
         # not existing missing values
         # TODO(jreback) what if we have a non-na float as a fill value?
-        if allow_fill and notna(fill_value):
+        # NaN with uncertainties is scalar but does not register as `isna`,
+        # so use fact that NaN != NaN
+        if allow_fill and notna(fill_value) and fill_value == fill_value:
```

Comment:
how would you even get here with uncertainties?  is your EA subclassing BaseMaskedArray?

Reply:
The EA itself is a very vanilla EA.  The test suite instantiates the EA in myriad ways for test purposes.  In this case the EA gets instantiated as a MaskedArray (which derives from BaseMaskedArray) for testing MaskedArray functionality.  I think the test suite has every right to rigorously instantiate EAs in every way imaginable, which it does.

Reply:
im still confused here.  the EA i see in https://github.com/hgrecco/pint-pandas/pull/140 subclasses ExtensionArray but not BaseMaskedArray, so i don't see how the code change here would affect it.

Reply:
OK...I looked more deeply into this when I get back the computer where I did original work and didn't upload some notes to the cloud.  I'm pretty sure it's due to some conditions created when trying to merge two DataFrames that I dummied up in my notes.  I'll create a test case that can go into the Pandas test suite.  It won't depend on uncertainties, but if and when used with uncertainties, it should trigger the path in question.  Should be done by Monday.

Reply:
Well, that was a lot more challenging to reconstruct than I expected, but I re-created the condition.  TL;DR: now that I use pd.NA as my NA value for my particular EA, this change that caught your eye is indeed unnecessary and I can revert it.

But, since you asked, here's the path I took initially.  Starting with `PintType` we define this:

```
    @property
    def na_value(self):
        if HAS_UNCERTAINTIES:
            return self.ureg.Quantity(_ufloat_nan, self.units)
        else:
            return self.ureg.Quantity(np.nan, self.units)
```

The idea (before settling on pd.NA, which works without this particular change) was to use a NaN to which we could cast any numerical type (we cannot downcast uncertainties to `float`).

In the test code we set up `data_missing` and `data_missing_for_sorting` to use `ufloat(np.nan, 0)` as the NA value where appropriate.   The test case `test_groupby_agg_extension` uses `data_for_grouping`, which is called without `numeric_dtype` as `None`, and this is where we start going off the rails. 

Here's the `data_for_grouping` code:

```
@pytest.fixture
def data_for_grouping(numeric_dtype, USE_UNCERTAINTIES):
    a = 1.0
    b = 2.0**32 + 1
    c = 2.0**32 + 10
    if USE_UNCERTAINTIES:
        a = a + ufloat(0, 0)
        b = b + ufloat(0, 0)
        c = c + ufloat(0, 0)
        _n = _ufloat_nan
        numeric_dtype = None
    elif numeric_dtype:
        numeric_dtype = dtypemap.get(numeric_dtype, numeric_dtype)
        _n = np.nan
    else:
        _n = pd.NA
    return PintArray.from_1darray_quantity(
        ureg.Quantity(pd.array([b, b, _n, _n, a, a, b, c], dtype=numeric_dtype), ureg.m)
    )
```

The way the test code is run, because `numeric_dtype` is `None` we get:

```
(Pdb) data_for_grouping
<PintArray>
[4294967297.0, 4294967297.0, <NA>, <NA>, 1.0, 1.0, 4294967297.0, 4294967306.0]
Length: 8, dtype: pint[meter]
```

When we execute `expected = df.iloc[[0, 2, 4, 7]]` there comes a point where we come here (`/Users/michael/Documents/GitHub/MichaelTiemannOSC/pandas-dev/pandas/core/internals/managers.py(646)<listcomp>()`):

```
644  	        else:
645  	            new_blocks = [
646  ->	                blk.take_nd(
647  	                    indexer,
648  	                    axis=1,
649  	                    fill_value=(
650  	                        fill_value if fill_value is not None else blk.fill_value
651  	                    ),
```

With `fill_value` as None, we inherit the fill value from the PintArray, which is the uncertain NaN.  Here's the stack trace down from there:

```
  pandas/core/internals/managers.py(645)reindex_indexer()
-> new_blocks = [
  pandas/core/internals/managers.py(646)<listcomp>()
-> blk.take_nd(
  pandas/core/internals/blocks.py(975)take_nd()
-> new_values = algos.take_nd(
  pandas/core/array_algos/take.py(115)take_nd()
-> return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill)
  pint-pandas/pint_pandas/pint_array.py(523)take()
-> result = take(data, indices, fill_value=fill_value, allow_fill=allow_fill)
  pandas/core/algorithms.py(1307)take()
-> result = take_nd(
  pandas/core/array_algos/take.py(115)take_nd()
-> return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill)
> pandas/core/arrays/masked.py(879)take()
-> result[fill_mask] = fill_value
```

In this particular case, because none of the indexes are -1, no filling is going to happen.  But should we have things to fill, they'd be filled with the uncertain nan.  And because this code really doesn't want to fill add new NaNs to the array, I put this extra defensive code in.  And that's how we hit this condition.

I'm going to take this extraneous test out (`fill_value == fill_value`).  Does this answer your question?